packages/babel-parser/src/parser/lval.js

```diff
@@ -376,64 +376,66 @@ export default class LValParser extends NodeUtils {
+   *
+   * @param {Expression} expr The given node
+   * @param {string} contextDescription The auxiliary context information printed when error is thrown
+   * @param {BindingTypes} [bindingType=BIND_NONE] The desired binding type. If the given node is an identifier and `bindingType` is not
+                                                   BIND_NONE, `checkLVal` will register binding to the parser scope
+                                                   See also src/util/scopeflags.js
+   * @param {?Set<string>} checkClashes An optional string set to check if an identifier name is included. `checkLVal` will add checked
+                                        identifier name to `checkClashes` It is used in tracking duplicates in function parameter lists. If
+                                        it is nullish, `checkLVal` will skip duplicate checks
+   * @param {boolean} [disallowLetBinding] Whether an identifier named "let" is allowed
+   * @param {boolean} [strictModeChanged=false] Whether an identifier have been parsed in a sloppy context but should be reinterpreted as
+                                                strict-mode. e.g. `(arguments) => { "use strict "}`
+   * @memberof LValParser
+   */
   checkLVal(
     expr: Expression,
-    bindingType: BindingTypes = BIND_NONE,
-    checkClashes: ?{ [key: string]: boolean },
     contextDescription: string,
+    bindingType: BindingTypes = BIND_NONE,
+    checkClashes: ?Set<string>,
     disallowLetBinding?: boolean,
     strictModeChanged?: boolean = false,
   ): void {
     switch (expr.type) {
-      case "Identifier":
+      case "Identifier": {
+        const { name } = expr;
         if (
           this.state.strict &&
           // "Global" reserved words have already been checked by parseIdentifier,
           // unless they have been found in the id or parameters of a strict-mode
           // function in a sloppy context.
           (strictModeChanged
-            ? isStrictBindReservedWord(expr.name, this.inModule)
-            : isStrictBindOnlyReservedWord(expr.name))
+            ? isStrictBindReservedWord(name, this.inModule)
+            : isStrictBindOnlyReservedWord(name))
         ) {
           this.raise(
             expr.start,
             bindingType === BIND_NONE
               ? Errors.StrictEvalArguments
               : Errors.StrictEvalArgumentsBinding,
-            expr.name,
+            name,
           );
         }
 
         if (checkClashes) {
-          // we need to prefix this with an underscore for the cases where we have a key of
-          // `__proto__`. there's a bug in old V8 where the following wouldn't work:
-          //
-          //   > var obj = Object.create(null);
-          //   undefined
-          //   > obj.__proto__
-          //   null
-          //   > obj.__proto__ = true;
-          //   true
-          //   > obj.__proto__
-          //   null
-          const key = `_${expr.name}`;
```

Comment:
Replaced `checkClashes` with `WeakSet` so we don't have to deal with old V8 bug.